gcm/gcm_connector.go

```diff
@@ -35,45 +40,74 @@ func NewGCMConnector(router server.Router, prefix string, gcmAPIKey string) (*GC
 		return nil, err
 	}
 
+	//TODO Cosmin: check with dev-team the number of GCM workers, below
 	gcm := &GCMConnector{
 		router:            router,
 		kvStore:           kvStore,
 		prefix:            prefix,
 		channelFromRouter: make(chan server.MsgAndRoute, 1000),
 		stopChan:          make(chan bool, 1),
 		sender:            &gcm.Sender{ApiKey: gcmAPIKey},
+		workersNumber:     runtime.GOMAXPROCS(0),
 	}
 
 	return gcm, nil
 }
 
-// Start opens the connector and awaits for messages from router to be forwarded to gcm until the stop signal is emitted
+// Start opens the connector, start more goroutines / workers to handle messages coming from the router
 func (conn *GCMConnector) Start() error {
 	broadcastRoute := server.NewRoute(removeTrailingSlash(conn.prefix)+"/broadcast", conn.channelFromRouter, "gcm_connector", "gcm_connector")
 	conn.router.Subscribe(broadcastRoute)
 	go func() {
+		//TODO Cosmin: should loadSubscriptions() be taken out of this goroutine, and executed before ?
+		// even if startup-time is longer, the routes are guaranteed to be there right after Start() returns
 		conn.loadSubscriptions()
 
-		for {
-			select {
-			case msg := <-conn.channelFromRouter:
-				if string(msg.Message.Path) == removeTrailingSlash(conn.prefix)+"/broadcast" {
-					go conn.broadcastMessage(msg)
-				} else {
-					go conn.sendMessage(msg)
-				}
-			case <-conn.stopChan:
-				return
-			}
+
+		protocol.Debug("number of GCM workers: %v", conn.workersNumber)
+		for i := 1; i <= conn.workersNumber; i++ {
+			protocol.Debug("starting GCM worker %v", i)
+			go conn.loopSendOrBroadcastMessage()
 		}
+		conn.waitGroup.Add(conn.workersNumber)
 	}()
 	return nil
 }
 
+// Stop signals the closing of GCMConnector
+func (conn *GCMConnector) Stop() error {
+	protocol.Debug("GCM Stop()")
+	close(conn.stopChan)
+	conn.waitGroup.Wait()
+	return nil
+}
+
+// Check returns nil if health-check succeeds, or an error if health-check fails
 func (conn *GCMConnector) Check() error {
 	return nil
 }
 
+// loopSendOrBroadcastMessage awaits in a loop for messages from router to be forwarded to GCM,
+// until the stop-channel is closed
+func (conn *GCMConnector) loopSendOrBroadcastMessage() {
+	for {
+		select {
+		case msg := <-conn.channelFromRouter:
+			if string(msg.Message.Path) == removeTrailingSlash(conn.prefix)+"/broadcast" {
+				go conn.broadcastMessage(msg)
+			} else {
+				go conn.sendMessage(msg)
+			}
+		case _, opened := <-conn.stopChan:
```

Comment:
There is no need for variables here. If a value is received, no matter what value, then we close.
Closing a channel send a `nil` to all listeners.